src/core/manager.c

```diff
@@ -2748,6 +2801,142 @@ static int manager_dispatch_notify_fd(sd_event_source *source, int fd, uint32_t
+                 * we won't take notification messages anymore, but that's still better than busy looping:
+                 * being woken up over and over again, but being unable to actually read the message from the
+                 * socket. */
+                return log_error_errno(n, "Failed to receive notification message for PAM pid: %m");
+
+        _cleanup_(pidref_done) PidRef parent_pidref = PIDREF_NULL, pam_pidref = PIDREF_NULL;
+
+        int *fds = CMSG_TYPED_DATA(cmsg_find(&msghdr, SOL_SOCKET, SCM_RIGHTS, CMSG_LEN(sizeof(int) * 2)), int);
+        if (fds) {
+                r = pidref_set_pidfd_consume(&parent_pidref, fds[0]);
+                if (r < 0) {
+                        log_warning_errno(r, "Received invalid main pid of PAM process, ignoring: %m");
+                        safe_close(fds[1]);
+                        return 0;
+                }
+
+                r = pidref_set_pidfd_consume(&pam_pidref, fds[1]);
+                if (r < 0) {
+                        log_warning_errno(r, "Received invalid PAM process for main pid ("PID_FMT"), ignoring: %m", parent_pidref.pid);
+                        return 0;
+                }
+        }
+
+        if (!fds) {
+                if ((size_t) n >= sizeof(buf) || (msghdr.msg_flags & MSG_TRUNC)) {
+                        log_warning("Received notify message exceeded maximum size. Ignoring.");
+                        return 0;
+                }
+
+                /* As extra safety check, let's make sure the string we get doesn't contain embedded NUL bytes.
+                 * We permit one trailing NUL byte in the message, but don't expect it. */
+                if (n > 1 && memchr(buf, 0, n-1)) {
+                        log_warning("Received notify message with embedded NUL bytes. Ignoring.");
+                        return 0;
+                }
+                /* Make sure it's NUL-terminated, then parse it to obtain the tags list. */
+                buf[n] = 0;
+
+                _cleanup_strv_free_ char **pidstrs = strv_split_newlines(buf);
+                if (!pidstrs) {
+                        log_oom();
+                        return 0;
+                }
+
+                if (strv_length(pidstrs) != 2) {
+                        log_warning("Received invalid pid string. Ignoring.");
+                        return 0;
+                }
+
+                r = pidref_set_pidstr(&parent_pidref, pidstrs[0]);
+                if (r < 0) {
+                        log_warning_errno(r, "Received invalid main pid (%s) of PAM process, ignoring: %m", pidstrs[0]);
+                        return 0;
+                }
+
+                r = pidref_set_pidstr(&pam_pidref, pidstrs[1]);
+                if (r < 0) {
+                        log_warning_errno(r, "Received invalid PAM process pid (%s) for main pid ("PID_FMT"), ignoring: %m",
+                                          pidstrs[1], parent_pidref.pid);
+                        return 0;
```

Comment:
Also here… This is essentially internal communication. If we get invalid messages, things are seriously wrong, we can't monitor the child precesses, so just kill the service?